pkg/controller/common/helpers.go

```diff
@@ -1263,3 +1264,68 @@ func RequiresRebuild(oldMC, newMC *mcfgv1.MachineConfig) bool {
 		!reflect.DeepEqual(oldMC.Spec.Extensions, newMC.Spec.Extensions) ||
 		!reflect.DeepEqual(oldMC.Spec.KernelArguments, newMC.Spec.KernelArguments)
 }
+
+// GetUpdatedMachines filters the provided nodes to return the nodes whose
+// current config matches its desired config and target config in the
+// associated MCP and has the "done" flag set.
+func GetUpdatedMachines(pool *mcfgv1.MachineConfigPool, nodes []*corev1.Node, mosc *mcfgv1.MachineOSConfig, mosb *mcfgv1.MachineOSBuild, layered bool) []*corev1.Node {
+	var updated []*corev1.Node
+	for _, node := range nodes {
+		lns := NewLayeredNodeState(node)
+		if lns.IsDone(pool, layered, mosc, mosb) {
+			updated = append(updated, node)
+		}
+	}
+	return updated
+}
+
+// GetDegradedMachines filters the provided nodes to return the nodes that
+// are considered in a degraded state.
+func GetDegradedMachines(nodes []*corev1.Node) []*corev1.Node {
+	var degraded []*corev1.Node
+	for _, node := range nodes {
+		if node.Annotations == nil {
+			continue
+		}
+		dconfig, ok := node.Annotations[daemonconsts.DesiredMachineConfigAnnotationKey]
+		if !ok || dconfig == "" {
```

Comment:
random question, if there is no desired config, does that mean the node is very new? Why do we need to explicitly skip over these?